### PR TITLE
Generate PyTorch artifacts

### PR DESCRIPTION
**Pull request by mariecwhite**
Adds functionality to generate PyTorch artifacts and updates artifacts.

### CODE REVIEW

File: .github/workflows/run_comparative_benchmark.yml [hunk 1]
```diff
@@ -13,6 +13,7 @@ on:
     # Scheduled to run at 09:00 UTC and 21:00 UTC.
     - cron: '0 09,21 * * *'
   workflow_dispatch:
+  pull_request:
```
Comment (mariecwhite) [inline .github/workflows/run_comparative_benchmark.yml:16]: Reminder to remove.

File: common_benchmark_suite/openxla/benchmark/comparative_suite/pt/model_definitions.py [hunk 2]
```diff
@@ -102,3 +143,11 @@
...
     template=RESNET50_FP16_PT_3X224X224XF16_BATCH_TEMPLATE,
     batch_sizes=[1, 8, 64, 128, 256, 2048])
+
+ALL_MODELS = list(
+    itertools.chain(
+        BERT_LARGE_FP32_PT_384XI32_BATCHES.values(),
+        BERT_LARGE_FP16_PT_384XI32_BATCHES.values(),
+        RESNET50_FP32_PT_3X224X224XF32_BATCHES.values(),
+        RESNET50_FP16_PT_3X224X224XF16_BATCHES.values(),
+    ))
```
Comment (pzread) [inline common_benchmark_suite/openxla/benchmark/comparative_suite/pt/model_definitions.py:153]: nit: Missing a newline

File: common_benchmark_suite/openxla/benchmark/comparative_suite/pt/scripts/generate_model_artifacts.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+# Add openxla dir to the search path.
+sys.path.insert(0, str(pathlib.Path(__file__).parents[5]))
+from openxla.benchmark import def_types
+from openxla.benchmark.comparative_suite.pt import model_definitions
+from openxla.benchmark.models import model_interfaces, utils
+
+
+def _generate_artifacts(model: def_types.Model, save_dir: pathlib.Path):
+  model_dir = save_dir.joinpath(model.name)
```
Comment (pzread) [inline common_benchmark_suite/openxla/benchmark/comparative_suite/pt/scripts/generate_model_artifacts.py:27]: nit:
```suggestion
  model_dir = save_dir / model.name
```

File: common_benchmark_suite/openxla/benchmark/comparative_suite/pt/scripts/generate_model_artifacts.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  try:
+    # Remove all gradient info from models and tensors since these models are inference only.
+    with torch.no_grad():
+      model_module = importlib.import_module(model.model_impl.module_path)
+      model_obj: model_interfaces.InferenceModel = model_module.create_model(
+          **model.model_parameters)
+      if model_obj.import_on_gpu and not torch.cuda.is_available():
+        raise RuntimeError("Model can only be exported on CUDA.")
+
+      model_obj = model_obj.to(dtype=model_obj.dtype)
```
Comment (pzread) [inline common_benchmark_suite/openxla/benchmark/comparative_suite/pt/scripts/generate_model_artifacts.py:40]: I think the model has been converted to the target type, so I guess we don't need to convert again here.
https://github.com/openxla/openxla-benchmark/blob/main/common_benchmark_suite/openxla/benchmark/models/pt/bert/bert_model.py#L95

File: common_benchmark_suite/openxla/benchmark/comparative_suite/pt/scripts/generate_model_artifacts.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+      if model_obj.import_with_fx:
+        mlir_data = import_torch_module_with_fx(
+            model_obj, inputs, torch_mlir.OutputType.LINALG_ON_TENSORS)
+      else:
+        graph = torch.jit.trace(model_obj, inputs)
+        mlir_data = import_torch_module(graph, inputs,
+                                        torch_mlir.OutputType.LINALG_ON_TENSORS)
+
+      # Save mlir.
+      mlir_path = model_dir.joinpath("linalg.mlirbc")
```
Comment (pzread) [inline common_benchmark_suite/openxla/benchmark/comparative_suite/pt/scripts/generate_model_artifacts.py:55]: nit:
```suggestion
      mlir_path = model_dir / "linalg.mlirbc"
```

File: common_benchmark_suite/openxla/benchmark/comparative_suite/pt/scripts/generate_model_artifacts.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+      print(f"Saving mlir to {mlir_path}")
+      with open(mlir_path, "wb") as f:
+        f.write(mlir_data)
```
Comment (pzread) [inline common_benchmark_suite/openxla/benchmark/comparative_suite/pt/scripts/generate_model_artifacts.py:58]: nit:
```suggestion
      mlir_path.write_bytes(mlir_data)
```

File: common_benchmark_suite/openxla/benchmark/comparative_suite/pt/scripts/generate_model_artifacts.sh [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+# Runs `generate_model_artifacts.py` on all registered PyTorch models and saves
+# artifacts into the directory `/tmp/pt_models_<torch-mlir-version>_<timestamp>`.
+#
+# Once complete. please upload the output directory to
+# `gs://iree-model-artifacts/pt`, preserving directory name.
+#
+# Usage:
+#     bash generate_saved_models.sh
+#
+# Requires python-3.11 and above and python-venv.
```
Comment (pzread) [inline common_benchmark_suite/openxla/benchmark/comparative_suite/pt/scripts/generate_model_artifacts.sh:18]: I thought PyTorch would need to run on Python 3.10? Or it's just for PyTorch Inductor

File: common_benchmark_suite/openxla/benchmark/def_types.py [hunk 8]
```diff
@@ -42,7 +42,8 @@ class ModelArtifactType(Enum):
   TF_SAVEDMODEL_V1 = "tf_savedmodel_v1"
   TF_SAVEDMODEL_V2 = "tf_savedmodel_v2"
   XLA_HLO_DUMP = "xla_hlo_dump"
-  STABLEHLO = "stablehlo"
+  STABLEHLO_MLIR = "stablehlo_mlir"
+  LINALG_MLIR = "linalg_mlir"
```
Comment (pzread) [inline common_benchmark_suite/openxla/benchmark/def_types.py:46]: Same as XLA HLO, it's a little bit questionable if Linalg MLIR can be considered as a stable format. I guess it's fine for now.

File: common_benchmark_suite/openxla/benchmark/models/pt/resnet/resnet_model.py [hunk 9]
```diff
@@ -73,6 +76,8 @@ def postprocess(self, outputs: Tuple[Any, ...]) -> Tuple[Any, ...]:
 def create_model(batch_size: int = 1,
                  data_type: str = "fp32",
                  model_name: str = "torchvision/resnet50",
+                 import_on_gpu: bool = False,
+                 import_with_fx: bool = True,
```
Comment (pzread) [inline common_benchmark_suite/openxla/benchmark/models/pt/resnet/resnet_model.py:80]: The method comment below also needs to be updated.

File: common_benchmark_suite/openxla/benchmark/models/pt/bert/bert_model.py [hunk 10]
```diff
@@ -71,6 +77,8 @@ def create_model(batch_size: int = 1,
                  seq_len: int = 384,
                  data_type: str = "fp32",
                  model_name: str = "bert-large-uncased",
+                 import_on_gpu: bool = False,
+                 import_with_fx: bool = True,
```
Comment (pzread) [inline common_benchmark_suite/openxla/benchmark/models/pt/bert/bert_model.py:81]: The method comment below also needs to be updated.

**Review by pzread — APPROVED** [review]

File: common_benchmark_suite/openxla/benchmark/comparative_suite/pt/model_definitions.py (hunk 2 above)
Comment (mariecwhite) [reply common_benchmark_suite/openxla/benchmark/comparative_suite/pt/model_definitions.py:153]: Added

File: common_benchmark_suite/openxla/benchmark/comparative_suite/pt/scripts/generate_model_artifacts.py (hunk 3 above)
Comment (mariecwhite) [reply common_benchmark_suite/openxla/benchmark/comparative_suite/pt/scripts/generate_model_artifacts.py:27]: Changed

File: common_benchmark_suite/openxla/benchmark/comparative_suite/pt/scripts/generate_model_artifacts.py (hunk 4 above)
Comment (mariecwhite) [reply common_benchmark_suite/openxla/benchmark/comparative_suite/pt/scripts/generate_model_artifacts.py:40]: Ah didn't see that, thanks!

File: common_benchmark_suite/openxla/benchmark/comparative_suite/pt/scripts/generate_model_artifacts.py (hunk 5 above)
Comment (mariecwhite) [reply common_benchmark_suite/openxla/benchmark/comparative_suite/pt/scripts/generate_model_artifacts.py:55]: Changed

File: common_benchmark_suite/openxla/benchmark/comparative_suite/pt/scripts/generate_model_artifacts.py (hunk 6 above)
Comment (mariecwhite) [reply common_benchmark_suite/openxla/benchmark/comparative_suite/pt/scripts/generate_model_artifacts.py:58]: Changed

File: common_benchmark_suite/openxla/benchmark/comparative_suite/pt/scripts/generate_model_artifacts.sh (hunk 7 above)
Comment (mariecwhite) [reply common_benchmark_suite/openxla/benchmark/comparative_suite/pt/scripts/generate_model_artifacts.sh:18]: It's `torch-mlir` that needs 3.11 and above.

File: common_benchmark_suite/openxla/benchmark/def_types.py (hunk 8 above)
Comment (mariecwhite) [reply common_benchmark_suite/openxla/benchmark/def_types.py:46]: We've been running the same linalg mlir files in IREE for the last 6 months without any issues. So far so stable.

File: common_benchmark_suite/openxla/benchmark/models/pt/resnet/resnet_model.py (hunk 9 above)
Comment (mariecwhite) [reply common_benchmark_suite/openxla/benchmark/models/pt/resnet/resnet_model.py:80]: Updated

File: common_benchmark_suite/openxla/benchmark/models/pt/bert/bert_model.py (hunk 10 above)
Comment (mariecwhite) [reply common_benchmark_suite/openxla/benchmark/models/pt/bert/bert_model.py:81]: Updated

**Pull request merged**